packages/cli/src/commands/deploy/action.ts

```diff
@@ -0,0 +1,63 @@
+import { TxParams } from '@openzeppelin/upgrades';
+import { compile } from '../../models/compiler/Compiler';
+import { fromContractFullName } from '../../utils/naming';
+import ConfigManager from '../../models/config/ConfigManager';
+import NetworkController from '../../models/network/NetworkController';
+import stdout from '../../utils/stdout';
+import { MethodArgType } from '../../prompts/prompt';
+import { parseMultipleArgs } from '../../utils/input';
+import NetworkFile from '../../models/files/NetworkFile';
+import { createAction } from '../create';
+
+import { AbortAction } from '../../register-command';
+
+interface Options {
+  network: string;
+  skipCompile?: boolean;
+  upgradeable?: boolean;
+
+  // The following are not available as CLI flags, and they are only used in tests.
+  networkFile?: NetworkFile;
+  txParams?: TxParams;
+}
+
+export async function preAction(
+  contractName: string | undefined,
+  deployArgs: string[],
+  options: Options,
+): Promise<void> {
+  if (!options.skipCompile) {
+    await compile();
+  }
+
+  if (options.upgradeable === true) {
+    throw new AbortAction(async () => {
```

Comment:
You're right. I intuitively went for an exception because I was thinking it would be necessary to jump over several levels of function calls but it isn't.

Reply:
> I wouldn't say exceptions necessarily mean that something went wrong
I agree. It depends on language and code style, for example in Python developers throw exceptions all the time and use them as a way to return from a function instead of returning `null`.